fdbclient/S3Client.actor.cpp

```diff
@@ -74,27 +77,212 @@ Reference<S3BlobStoreEndpoint> getEndpoint(const std::string& s3url,
 	return endpoint;
 }
 
+// State for a part of a multipart upload.
+struct PartState {
+	int partNumber;
+	std::string etag;
+	int64_t offset;
+	int64_t size;
+	std::string md5; // MD5 from S3 metadata if available
+	bool completed = false;
+};
+
+// Config for a part of a multipart upload.
+// TODO: Make this settable via knobs.
+struct PartConfig {
+	int64_t partSizeBytes = 8 * 1024 * 1024; // Default 16MB
+	// Max file size seems to be 32MB or so.... Lets have one thread only for the moment.
+	int maxConcurrent = 4;
+	int maxRetries = 3;
+	int retryDelayMs = 1000;
+};
+
+// Upload a part of a multipart upload with retry logic.
+ACTOR static Future<Void> uploadPartWithRetry(Reference<S3BlobStoreEndpoint> endpoint,
+                                              std::string bucket,
+                                              std::string objectName,
+                                              std::string uploadID,
+                                              PartState* part,
+                                              std::string partData,
+                                              int maxRetries,
+                                              int retryDelayMs) {
+	state int attempts = 0;
+	// Create packet queue for the part data
+	state UnsentPacketQueue packets;
+
+	while (attempts < maxRetries) {
+		try {
+			// Verify MD5 before upload
+			std::string calculatedMD5 = HTTP::computeMD5Sum(partData);
+			if (part->md5 != calculatedMD5) {
+				TraceEvent(SevWarn, "S3ClientUploadPartMD5Mismatch")
+				    .detail("Bucket", bucket)
+				    .detail("Object", objectName)
+				    .detail("Part", part->partNumber)
+				    .detail("Expected", part->md5)
+				    .detail("Calculated", calculatedMD5);
+				throw io_error();
+			}
+			PacketWriter pw(packets.getWriteBuffer(partData.size()), nullptr, Unversioned());
+			pw.serializeBytes(partData);
+
+			state std::string etag = wait(endpoint->uploadPart(
+			    bucket, objectName, uploadID, part->partNumber, &packets, partData.size(), calculatedMD5));
+			part->etag = etag;
+			part->completed = true;
+			return Void();
+		} catch (Error& e) {
+			attempts++;
+			if (attempts >= maxRetries) {
+				throw;
```

Comment:
What if the MD5 is consistently failed? We may want to add a TODO(BulkLoad) here. We need an error indicating an unretriable error happens.

Reply:
Added.